src/main/java/com/google/sps/Utility.java

```diff
@@ -108,55 +110,66 @@ public static String graphToJson(
    * @param original the original graph
    * @param curr the current (most recently-requested) graph (requires that original != curr)
    * @param mutationNum number of mutations to apply
-   * @param multiMutList multi-mutation list
+   * @param multiMutList multi-mutation list builder
    * @throws IllegalArgumentException if original and current graph refer to the same object
    * @return the resulting data graph, null if the mutation number was too small, and the final
-   *     graph if the mutation number was too big
+   *     graph if the mutation number was too big. May also modify multiMutList, replacing some
+   *     mutations with their deduplicated versions.
    */
   public static DataGraph getGraphAtMutationNumber(
-      DataGraph original, DataGraph curr, int mutationNum, List<MultiMutation> multiMutList)
+      DataGraph original, DataGraph curr, int mutationNum, MutationList.Builder mutationsList)
       throws IllegalArgumentException {
     Preconditions.checkArgument(
         original != curr, "The current graph and the original graph refer to the same object");
 
+    List<MultiMutation> multiMutList = mutationsList.getMutationList();
+
     if (mutationNum < -1) {
       return null;
     } else if (mutationNum > multiMutList.size()) {
       mutationNum = multiMutList.size() - 1;
     }
+    // If the requested graph is sequentially before the current graph but is closer to
+    // the initial graph than to the current graph, go forward from the initial graph rather
+    // than going back from the current graph
+    if (curr.numMutations() > mutationNum && (curr.numMutations() - mutationNum) > mutationNum) {
```

Comment:
maybe put this in the else block on line 158 and just keep the `(curr.numMutations() - mutationNum) > mutationNum` part?